keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2368 @@
+makes it virtually agnostic of the hardware architecture. The OCI runtime-spec
+has supported Intel RDT for a while already (see the
+[OCI runtime-spec][oci-runtime-rdt]). Other hardware vendors have comparable
+technologies which use the same [resctrl interface][linux-resctrl].
+
+The Linux Block IO controller parameters depend very heavily on the underlying
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes has been
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients. Following this model also
+provides a possible framework for the future improvements, for instance
+enabling class-based network or memory type prioritization of applications.
+
+At the time of writing (May 2023), there is no mechanism in Kubernetes to use
+these types of resources. CRI-O and containerd runtimes have support for RDT
+and blockio classes, but rely on Pod annotations as the user interface, with an
+inferred expectation that people set the equivalent nonstandard and private
+annotations on a Kubernetes Pod. The goal of this KEP is to get these types of
+resources first class citizens and properly supported in Kubernetes, providing
+visibility, a well-defined user interface, and permission controls.
+
+Two types of QoS-class resources are identified, container-level and pod-level
+QoS-class resources. Container-level resources enable QoS on per-container
+granularity, for example container-level cgroups in Linux or cache and memory
+bandwidth control technologies. Examples for pod-level QoS include e.g.
+pod-level cgroups or network QoS that cannot support per-container granularity.
+
+### Goals
+
+- Make it possible to request QoS-class resources from the PodSpec
+  - Container-level QoS-class resources
+  - Pod-level (sandbox-level) QoS-class resources
+- Make it simple to implement new types QoS-class resource
+- Make QoS-class resources opaque (as possible) to Kubernetes
+- Support automatic discovery of the available QoS-class resources
+- Support per-node status/capacity of QoS-class resources
+- Access control ([future work](#future-work))
+- Support updating QoS-class resource assignment of running containers
+  ([future work](#in-place-pod-vertical-scaling))
+
+### Non-Goals
+
+- Interface or mechanism for configuring the QoS-class resources (responsibility of
+  the container runtime).
+- Specifying available (QoS-class) resource types or their detailed behavior
+- API changes to support updating Pod-level (sandbox-level) QoS-class resource
+  assignment of running pods (will be subject to a separate KEP)
+- Per runtime class QoS (needs a separate KEP if this is desired, as it must
```

Comment:
This seems rather important, doesn't it?  I don't know how often a node has more than one runtime class, but it doesn't seem right to assume that all QoS classes are present on all runtimes?

Reply:
I think this is a much broader problem that just QoS resources and thus would deserve a separate KEP. Currently, there is not concept of per-runtime-class ("nested") structuring of resources. The same problem of resource being unavailable for some runtime class may be the case for other resources as well (thinking about vm-based/confidential/non-runc containers).

WDYT @thockin?

Reply:
@thockin we had discussion about runtimeclasses with @mikebrow in the context of how it is implemented in containerd several months ago. The conclusion was that nobody is ready to open that pandora box to have all resources nested/flavoured on per runtimeclass. because that would hit everything, image pulling (different runtime  classes in containerd can have different snapshotters/handlers of imaging layer), not all native, extended or device plugin resources are potentially compatible with all runtime handlers and so on. That's the reasons why this KEP is declaring it out of scope and keeping the "flat" per-node resource model. If at some point we want to have resources to be nested under the runtime classes for capacities/compatibilities tracking, it would definitely deserve separate huge KEP, as it would touch practically every bit of resource accounting in every component.

Reply:
reading into Tim's question a bit.. let's discuss what should happen with respect to node assignment/startup of a pod that has QoS class(es) known to not be presently supported on the runtime class implementation requested by the pod spec. That would be sufficient perhaps? For example, runtime default is runc/crun and supports this class list...

Reply:
@marquiz - I agree with "The same problem ... may be the case for other resources as well".  That doesn't make me happy, though.